File: 백준/Silver/1149. RGB거리/RGB거리.py
cost=[[0]*3 for _ in range(1001)] # i번째 집을 r,g,b 로 칠했을때 각각 비용 
save=[0]*1001 # 입력값 저장 

n=int(input())
for i in range(1,n+1): # 입력 받아 save 리스트에 저장 
    save[i]=list(map(int,input().split()))

for i in range(1,n+1):
    cost[i][0]=min(cost[i-1][1],cost[i-1][2])+save[i][0]
    cost[i][1]=min(cost[i-1][0],cost[i-1][2])+save[i][1]
    cost[i][2]=min(cost[i-1][0],cost[i-1][1])+save[i][2]

print(min(cost[n][0],cost[n][1],cost[n][2]))

File: 백준/Silver/1149. RGB거리/RGB거리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	
	static int[][] dp;
	static int[][] rgb;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		int n=Integer.parseInt(br.readLine());
		rgb=new int[n][3];
		dp=new int[n][3];
		String[] temp;
		
		for(int i=0;i<n;i++) {
			temp=br.readLine().split(" ");
			for(int j=0;j<3;j++) rgb[i][j]=Integer.parseInt(temp[j]);
		} // 그래프 생성 

		for(int i=0;i<3;i++) dp[0][i]=rgb[0][i]; // 1행 복사 
		
		for(int i=1;i<n;i++) {
			paint(i);
		}
		
		// 마지막 행에서 최솟값 찾기 
		int min=Integer.MAX_VALUE;
		for(int i=0;i<3;i++) {
			if(dp[n-1][i]<min) min=dp[n-1][i];
		}
		System.out.println(min);
	}

	static void paint(int row) { // row 행 dp 배열 채우기 
		for(int i=0;i<3;i++) {
			if(i==0) dp[row][i]=Math.min(dp[row-1][1],dp[row-1][2])+rgb[row][0];
			else if(i==1) dp[row][i]=Math.min(dp[row-1][0], dp[row-1][2])+rgb[row][1];
			else dp[row][i]=Math.min(dp[row-1][0], dp[row-1][1])+rgb[row][2];
		}
	}
}